crates/ruff_python_semantic/src/scope.rs

```diff
@@ -23,6 +23,8 @@ pub struct Scope<'a> {
     bindings: FxHashMap<&'a str, BindingId>,
     /// A map from binding ID to binding ID that it shadows.
     shadowed_bindings: HashMap<BindingId, BindingId, BuildNoHashHasher<BindingId>>,
+    /// A list of all names declared in the scope.
+    declared_symbols: Vec<&'a str>,
```

Comment:
This might be a bit of a heavy-handed change just to fix this one semantic issue... E.g., we could instead track `deleted_names`, and make `declares` something like `self.has(name) || self.deletions.contains(name)`.

Reply:
(We could also track `HashMap<&'a str, SymbolId>`, and make `bindings: FxHashMap<SymbolId, BindingId>`.)

Reply:
I'm having a hard time comprehending the difference between a `declared_symbol` vs a `binding` based on their names and comments alone 🤯

Reply:
I'm going to rewrite this to special-case deletions. I need to reconsider deletions more holistically.